devito/interfaces.py

```diff
@@ -0,0 +1,45 @@
+import numpy as np
+from sympy import IndexedBase
+
+
+class MatrixData(IndexedBase):
+    def __init__(self, name, shape, dtype):
+        self.name = name
+        self.shape = shape
+        self.dtype = dtype
+        self.pointer = None
+        self.initializer = None
+
+    def set_initializer(self, lambda_initializer):
+        assert(callable(lambda_initializer))
+        self.initializer = lambda_initializer
+
+    def initialize(self):
+        assert(self.initializer is not None)
+        self.initializer(self.data)
+
+    def _allocate_memory(self):
+        self.pointer = np.zeros(self.shape, self.dtype, order='C')
+
+    @property
+    def data(self):
+        if self.pointer is None:
+            self._allocate_memory()
+        return self.pointer
+
+
+class TimeData(MatrixData):
```

Comment:
Not yet. Still working on the whole bunch of changes needed to make it work. Will commit those in a separate branch once I am done.